protocol/protocol.go

```diff
@@ -136,3 +151,28 @@ func addFlow(name string, router *routerpkg.Router, messageTypes []string, stopp
 		}
 	})
 }
+
+func addOneTimeFlow(name string, router *routerpkg.Router, messageTypes []string, stopped *uint32,
+	stopChan chan error, flow func(route *routerpkg.Route) (routeClosed bool, err error)) {
+
+	route, err := router.AddIncomingRoute(messageTypes)
+	if err != nil {
+		panic(err)
+	}
+	defer func() {
+		err := router.RemoveRoute(messageTypes)
+		if err != nil {
+			panic(err)
+		}
+	}()
```

Comment:
It appears that you remove the route immediately, as the rest of the function is async.